### ISSUE TITLE
Cannot encrypt new files unless all files are decrypted first

### ISSUE THREAD

**Issue by dave-pollock** [sobolevn/git-secret#253]
I was surprised to hit this issue when trying out git-secret for the first time. I'm hoping I'm just doing something dumb. Basically I cannot encrypt new files added to git-secret unless all other files in the mapping.cfg file exist. In my case they may not exist as not all contributors will be given access to private keys, but I still want them to be able to encrypt new files.

What are the steps to reproduce this issue?
-------------------------------------------
```bash
echo "Test1" > secret/file1.txt
git secret add secret/file1.txt
1 item(s) added.
git secret hide -d
done. all 1 files are hidden.
echo "Test2" > secret/file2.txt
git secret add secret/file2.txt
1 item(s) added.
git secret hide -d
git-secret: abort: file not found: secret/file1.txt
```

What happens?
-------------
I am unable to encrypt file2.txt without first decrypting file1.txt

What were you expecting to happen?
----------------------------------
I expect to be able to encrypt files even if other files which have been added to the mapping.cfg do not exist. I'd prefer to perhaps just see a warning if a file does not exist, but for git-secret to continue attempting to encrypt all other files.

What versions of software are you using?
----------------------------------------
**Operating system:** (`uname -a`) …
Linux 92abc9c47df7 4.9.93-linuxkit-aufs #1 SMP Wed Jun 6 16:55:56 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**`git-secret` path:** (`which git-secret`) …
/usr/bin/git-secret

**`git-secret` version:** (`git secret --version`) …
0.2.4

**`git` version:** (`git --version`) …
git version 2.7.4

**Shell type and version:** (`$SHELL --version`) …
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)

**`gpg` version:** (`gpg --version`) …
gpg (GnuPG) 1.4.20

**Comment by joshrabinowitz** [comment]
Hm, I can see some logic why git-secret wouldn't only encrypt some of the files: The keychain might have changed since the last encryption, and if we don't re-encrypt all of the files, some files may require different keys than others to decrypt.  

@sobolevn @simbo1905 any thoughts?

**Comment by joshrabinowitz** [comment]
I'm planning to have git-secret issue a more explanatory error message.

**Comment by dave-pollock** [comment]
I think the use case where a collaborator needs to add an encrypted secret but does not have access to the private key to decrypt the existing secrets is really useful, and just improving the error message doesn't help this use case.

Perhaps a warning instead of an error, but with an improved message? Something like:
`warning: file not found: secret/file1.txt. Any changes to the keychain since this file was last encrypted have not been applied.`

Another possibility would be an additional flag which could tell git-secret to ignore missing files? 
`git secret hide --ignore-missing-files`

Yet another possibility would be adding the ability to specify exactly which file to encrypt?
`git secret hide -f secret/file2.txt`

I'd be happy to attempt a pull request for any of these options if you think any of them sounds acceptable.

Cheers!

**Comment by joshrabinowitz** [comment]
It would be good to have @sobolevn to chime in here. 

I can see it from both sides, but I think a situation where a user thinks they can decrypt all the files but cannot is pretty bad. 

Maybe this should remain a hard error, but the user can `-f` force it to succeed anyway?

**Comment by simbo1905** [comment]
I have run into this when renaming secrets. I have also often run into cases where I only want to encrypt a single file and have ended up with many files encrypted even when using the `-m` flag. The suggestion of being able to pass the file to encrypt to just encrypt a single file would appear to solve both cases.

**Comment by dave-pollock** [comment]
Hey guys,

I've had a crack at adding a -f flag to `hide` - see https://github.com/sobolevn/git-secret/pull/265. Any feedback appreciated.

Cheers.

**Comment by joshrabinowitz** [comment]
@dave-pollock I also took a crack at this and went a different way: #263 (there's a broken test, but otherwise it looks good and has been approved for merge).

I think the way I did it is preferable (because of the explanation worked up with @sobolevn in https://github.com/sobolevn/git-secret/blob/82ef82a272560a0812d367ca564eddb6f381f4f1/man/man1/git-secret-hide.1.ronn ).  It's not safe to allow users to encrypt a subset of the repo's files - it can create a situation where "some files can no longer be decrypted by users who should be (and would appear) able to decrypt them, and vice-versa."

**Comment by dave-pollock** [comment]
@joshrabinowitz fine with me as long as my use case is possible :)

I'll kill my PR.

**Comment by joshrabinowitz** [comment]
Closed, -F 'force mode' implemented in #263
